File: codes/breadth_first_search.jl
# Breadth First Search(BFS) of a binary tree

# Tree structure
struct TreeNode
	node
	left
	right
end

# function to build a tree
function build_tree(index=0)
	# full binary tree: (7-nodes)
	if index == 1
		#=
			     1
			   ╱   ╲
		      2     3
		    ╱  ╲
		   4     5
		 ╱  ╲
		6    7
		=#
		
		println("Rendering full binary tree with 7 nodes...")
		t1 = TreeNode(4, TreeNode(6, nothing, nothing), TreeNode(7, nothing, nothing))
		t2 = TreeNode(2, t1, TreeNode(5, nothing, nothing))
		t = TreeNode(1, t2, TreeNode(3, nothing, nothing))
	
	# complete binary tree: (10-nodes)
	elseif index == 2
		#=
			        1
			     ╱    ╲
		       2        3
		     ╱  ╲      ╱  ╲
		   4      5   6	   7
		 ╱  ╲	 ╱
		8    9  0
		=#
		
		println("Rendering complete binary tree with 10 nodes...")
		t1 = TreeNode(4, TreeNode(8, nothing, nothing), TreeNode(9, nothing, nothing))
		t2 = TreeNode(2, t1, TreeNode(5, TreeNode(0, nothing, nothing), nothing))
		t3 = TreeNode(3, TreeNode(6, nothing, nothing), TreeNode(7, nothing, nothing))
		t = TreeNode(1, t2, t3)

	# perfect binary tree: (15-nodes)
	else
		if index != 0
			println("Invalid Index. Used 0 by default...")
		end
		#= 
			         1
			   	  ╱     ╲
			   	 ╱       ╲
		       2           3
		    ╱   ╲        ╱   ╲
		  4      5      6	  7
		 ╱ ╲	╱ ╲    ╱ ╲   ╱  ╲
		8   9  0   1  2   3 4    5
		=#
		println("Rendering perfect binary tree with 15 nodes...")
		t1 = TreeNode(4, TreeNode(8, nothing, nothing), TreeNode(9, nothing, nothing))
		t2 = TreeNode(5, TreeNode(0, nothing, nothing), TreeNode(1, nothing, nothing))
		t3 = TreeNode(6, TreeNode(2, nothing, nothing), TreeNode(3, nothing, nothing))
		t4 = TreeNode(7, TreeNode(4, nothing, nothing), TreeNode(5, nothing, nothing))
		t5 = TreeNode(2, t1, t2)
		t6 = TreeNode(3, t3, t4)
		t = TreeNode(1, t5, t6)
	end
	return t
end


# Implementation of inorder traversal of binary tree
function inorder_traversal(tree, _path=nothing)
	"""
	 	function for the inorder traversal
	"""
	if _path == nothing
		_path = []
	end
	if tree != nothing
		inorder_traversal(tree.left, _path)
		push!(_path, tree.node)
		inorder_traversal(tree.right, _path)
	end
	return _path
end


# Breath first search implementation
function breath_first_search(root)
	"""
		function for the breath first search
	"""
	if root == nothing
		return []
	end

	queue = [root]
	bfs_result = []
	while isempty(queue) != true
		current = queue[1]
		deleteat!(queue, 1)
		push!(bfs_result, current.node)
		if current.left != nothing
			push!(queue, current.left)
		end
		if current.right != nothing
			push!(queue, current.right)
		end
	end
	return bfs_result;
end

# main function
function main()
	tree = build_tree(1);
	inorder_path = inorder_traversal(tree);
	bfs_path = breath_first_search(tree);
	println("Inorder Traversal > ", inorder_path);
	println("Breadth First Search > ", bfs_path);
end

main()

File: codes/breadth_first_search.R
rm(list=ls())
# Breadth first search of a binary tree

# Function to build binary tree
make_tree <- function(index=0) {
  if (index == 1) {
    "
    Binary tree - Full binary tree with 7 nodes
           1
         /   \
        2     3
       /  \
      4    5
     / \
    6   7
    "
	print("Rendering full binary tree with 7 nodes...")
    t1 <- list(node = 6, left = NULL, right = NULL)
    t2 <- list(node = 7, left = NULL, right = NULL)
    t3 <- list(node = 4, left = t1, right = t2)
    t4 <- list(node = 5, left = NULL, right = NULL)
    t5 <- list(node = 2, left = t3, right = t4)
    t6 <- list(node = 3, left = NULL, right = NULL)
    t <- list(node = 1, left = t5, right = t6)
    return (t)
  } 
  else if (index == 2) {
    "
    Binary tree - Complete binary tree with 10 nodes
             1
          /    \
         2       3
       /  \     / \
      4     5  6   7
     / \   /
    8   9 0
    "
	print("Rendering complete binary tree with 10 nodes...")
    t1 <- list(node = 8, left = NULL, right = NULL)
    t2 <- list(node = 9, left = NULL, right = NULL)
    t3 <- list(node = 0, left = NULL, right = NULL)
    t4 <- list(node = 4, left = t1, right = t2)
    t5 <- list(node = 5, left = t3, right = NULL)
    t6 <- list(node = 6, left = NULL, right = NULL)
    t7 <- list(node = 7, left = NULL, right = NULL)
    t8 <- list(node = 2, left = t4, right = t5)
    t9 <- list(node = 3, left = t6, right = t7)
    t <- list(node = 1, left = t8, right = t9)
    return (t)
  } 
  else {
    if (index != 0) {
      print("Index not available. Used 0 by default...")
    }
    "
    Binary tree - Perfect binary tree with 15 nodes
             1
          /     \
         2        3
       /  \       /  \
      4    5     6     7
     /\    /\    /\    /\
    8  9  0  1  2  3  4  5
    "
	print("Rendering perfect binary tree with 15 nodes...")
    t1 <- list(node = 8, left = NULL, right = NULL)
    t2 <- list(node = 9, left = NULL, right = NULL)
    t3 <- list(node = 0, left = NULL, right = NULL)
    t4 <- list(node = 1, left = NULL, right = NULL)
    t5 <- list(node = 2, left = NULL, right = NULL)
    t6 <- list(node = 3, left = NULL, right = NULL)
    t7 <- list(node = 4, left = NULL, right = NULL)
    t8 <- list(node = 5, left = NULL, right = NULL)
    
    t9 <- list(node= 4, left= t1, right= t2)
    t10 <- list(node= 5, left= t3, right= t4)
    t11 <- list(node= 6, left= t5, right= t6)
    t12 <- list(node= 7, left= t7, right= t8)
    t13 <- list(node= 2, left= t9, right= t10)
    t14 <- list(node= 3, left= t11, right= t12)
    t <- list(node= 1, left= t13, right= t14)
    return (t)
  }
}

# Queue implementation
# >>> Basic functions:(push, pop(:= poll), size, isEmpty)
Queue <- setRefClass(Class = "Queue",
                     fields = list(
                       data = "list"),
                     methods = list(
                       size = function() {
                         return(length(data))
                       },
                       push = function(x) {
                         data[[length(data) + 1]] <<- x
                       },
                       pop = function() {
                         if (size() == 0) return (NULL)
                         x <- data[[1]]
                         data[[1]] <<- NULL
                         return (x)
                       },
                       isEmpty = function() {
                         return (length(data) == 0)
                       }
                     ))

# Linked list implementation
LinkedList <-setRefClass(Class = "LinkedList",
                         fields = list(
                           data = "list"),
                         methods = list(
                           size = function() {
                             return (length(size))
                           },
                           append = function(x) {
                             data[[length(data)+1]] <<- x
                           },
                           pop = function() {
                             if(length(data) != 0) {
                               val <- data[[length(data)]]
                               data[length(data)] <<- NULL
                               return (val)
                             }
                           },
                           isEmpty = function () {
                             return (length(data) == 0)
                           },
                           getList = function() {
                             res <- c()
                             for (x in data) {
                               res[length(res) + 1] <- x
                             }
                             return (res)
                           }
                         ))

# Breadth First Search
bfs <- function(tree) {
  bfs_path <- c()
  queue <- Queue$new()
  queue$push(tree)
  
  while (! queue$isEmpty()) {
    current <- queue$pop()
    bfs_path[length(bfs_path) + 1] <- current$node

    if (! is.null(current$left)) {
      queue$push(current$left)
    }
    if (! is.null(current$right)) {
      queue$push(current$right)
    }
  }
  return (bfs_path)
}

# Inorder Traversal helper function
inorder_traversal_util <- function(tree, inorder_result) {
  if (! is.null(tree)) {
    inorder_traversal_util(tree$left, inorder_result)
    inorder_result$append(tree$node)
    inorder_traversal_util(tree$right, inorder_result)
  }
}

# Inorder Traversal
inorder_traversal <- function(tree) {
  inorder_result <- LinkedList$new()
  inorder_traversal_util(tree, inorder_result)
  
  return (inorder_result$getList())
}

tree <- make_tree(1)
inorder_path <- inorder_traversal(tree)
bfs_path <- bfs(tree)
cat("Inorder traversal > ", inorder_path)
cat("Breadth First Search > ", bfs_path)
